src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -190,135 +238,162 @@ public long ramBytesUsed() {
+                throw new FileNotFoundException("committed translog id can not be found (current ["
+                        + Strings.collectionToCommaDelimitedString(currentIds) + "], got [" + translogId + "]");
 
-    @Override
-    public void newTransientTranslog(long id) throws TranslogException {
-        rwl.writeLock().lock();
-        try {
-            assert this.trans == null;
-            long size = Long.MAX_VALUE;
-            Path location = null;
-            for (Path file : locations) {
-                long currentFree = Files.getFileStore(file).getUsableSpace();
-                if (currentFree < size) {
-                    size = currentFree;
-                    location = file;
-                } else if (currentFree == size && ThreadLocalRandom.current().nextBoolean()) {
-                    location = file;
+            }
+            lastCommitedTranslogId = translogId;
+            if (uncomittedTranslogs.isEmpty()) {
+                return;
+            }
+            while (uncomittedTranslogs.isEmpty() == false && uncomittedTranslogs.get(0).translogId() < translogId) {
+                FsChannelReader old = uncomittedTranslogs.remove(0);
+                logger.trace("removed [{}] from uncommitted translog list", old.translogId());
+                try {
+                    IOUtils.close(old);
+                } catch (IOException e) {
+                    logger.error("failed to closed old translog [{}] (committed id [{}])", e, old, translogId);
                 }
             }
-            this.trans = type.create(shardId, id, new InternalChannelReference(location.resolve(getPath(id)), StandardOpenOption.READ, StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW), transientBufferSize);
-        } catch (IOException e) {
-            throw new TranslogException(shardId, "failed to create new translog file", e);
-        } finally {
-            rwl.writeLock().unlock();
         }
     }
 
     @Override
-    public void makeTransientCurrent() throws IOException {
-        FsTranslogFile old;
-        rwl.writeLock().lock();
-        try {
-            assert this.trans != null;
-            old = current;
-            this.current = this.trans;
-            this.trans = null;
-        } finally {
-            rwl.writeLock().unlock();
+    public long newTranslog() throws TranslogException, IOException {
+        try (ReleasableLock _ = writeLock.acquire()) {
+            final FsTranslogFile newFile = createTranslogFile();
+            final FsTranslogFile old = current;
+            current = newFile;
+            newFile.reuse(old);
+            FsChannelImmutableReader reader = old.immutableReader();
+            uncomittedTranslogs.add(reader);
+            for (FsView view : outstandingViews) {
+                view.onNewTranslog(old.immutableReader(), current.reader());
+            }
+            IOUtils.close(old);
+            logger.trace("current translog set to [{}]", current.translogId());
```

Comment:
Needs to be protected by `logger.isTraceEnabled()` now that `translogId()` locks the readlock, either that, or grab it into a temporary variable before the logging and return statements.